### ISSUE TITLE
Make number of snippets shown in list view adjustable via `SnippetViewSet`

### ISSUE THREAD

**Issue by bmoe872** [wagtail/wagtail#7399]
### Is your proposal related to a problem?

It would be great to be able to adjust the number of snippets shown in the list page. Sometimes being able to see ALL the snippets is helpful. 

### Describe the solution you'd like

I think adjusting the code here:
https://github.com/wagtail/wagtail/blob/b8a86ac15de331654024f3b4590088f0b88626fc/wagtail/snippets/views/snippets.py#L145

To accept a setting would make this work. If the setting is not available we could default to 20

@gasman has made a better suggestion below. Stating that this could be added to the model similarly to how ModelAdmin handles it with a config object.  

### Describe alternatives you've considered

I thought about overwriting the view myself, but there's a lot there, and I think this could be better served through a setting in settings.py

**Comment by bmoe872** [comment]
I like this suggestion! I think that makes way more sense to do it on a per_model basis!

**Comment by lb-** [comment]
@gasman  would love your thoughts on this rough implementation.

* `register_snippet` becomes a potential 'partial' function so it can be used as currently documented and also used with an arg passed with a config class.
* However, this will fail if you call it as a function with only one arg (however, this use case is not in the documentation, only the decorator approach is documented).
* Not sure on naming do just went with `snippet_configuration` to avoid clashes with using `configuration`
* Went with the config page length attribute name of `list_per_page` as this is what ModelAdmin uses


**wagtail/snippets/models.py**
```python
from django.contrib.admin.utils import quote
from django.core import checks
from django.urls import reverse

from wagtail.admin.checks import check_panels_in_model
from wagtail.admin.models import get_object_usage


SNIPPET_MODELS = []


class DefaultSnippetConfig:
    list_per_page = 20


def get_snippet_models():
    return SNIPPET_MODELS


def _register_snippet(model, config=DefaultSnippetConfig):
    if model not in SNIPPET_MODELS:
        model.snippet_configuration = config
        model.get_usage = get_object_usage
        model.usage_url = get_snippet_usage_url
        SNIPPET_MODELS.append(model)
        SNIPPET_MODELS.sort(key=lambda x: x._meta.verbose_name)

        @checks.register('panels')
        def modeladmin_model_check(app_configs, **kwargs):
            errors = check_panels_in_model(model, 'snippets')
            return errors

    return model


def register_snippet(*args):

    if len(args) == 2:
        return _register_snippet(*args)

    if hasattr(args[0], 'objects'):
        model = args[0]
        return _register_snippet(model)

    def wrapper(model):
        config = args[0]
        return _register_snippet(model, config)

    return wrapper


def get_snippet_usage_url(self):
    return reverse('wagtailsnippets:usage', args=(
        self._meta.app_label, self._meta.model_name, quote(self.pk)))

```

**wagtail/snippets/views/snippets.py**
```
class ListView(TemplateView):
    # ...

    def get_context_data(self, **kwargs):
        # ....
        configuration = self.model.snippet_configuration # get configuration set on model

        paginator = Paginator(items, per_page=getattr(configuration, 'list_per_page', 20)) # data from config
        paginated_items = paginator.get_page(self.request.GET.get('p'))

        context.update({
```

**Comment by laymonage** [comment]
This will be implemented in #10241 through `SnippetViewSet`.

**Comment by gasman** [comment]
Completed in #10241
